sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala

```diff
@@ -285,6 +287,110 @@ case class ArrayTransform(
+    "than the second nullable element. Null elements will be placed at the end of the returned" +
+    "array. If the comparator function returns other values (including NULL)," +
+    "the query will fail and raise an error. By the default it will sort the array in " +
+    "ascending mode",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(5, 6, 1), (x,y) -> f(x,y));
+       [1,5,6]
+      > SELECT _FUNC_(array('bc', 'ab', 'dc'), (x, y) -> f(x,y));
+       ['dc', 'bc', 'ab']
+  """,
+  since = "3.0.0")
+case class ArraySort(
+    argument: Expression,
+    function: Expression)
+  extends ArrayBasedSimpleHigherOrderFunction with ArraySortLike with CodegenFallback {
+
+  def this(argument: Expression) = this(argument, Literal(true))
+
+  @transient lazy val argumentsType: DataType =
+    argument.dataType.asInstanceOf[ArrayType].elementType
+
+  override protected def arrayExpression: Expression = argument
+
+  override protected def nullOrder: NullOrder = NullOrder.Greatest
+
+  override def dataType: ArrayType = ArrayType(argumentsType, argument.nullable)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+
+    if (function.dataType == BooleanType) {
+      argument.dataType match {
+        case ArrayType(dt, _) if RowOrdering.isOrderable(dt) =>
+          TypeCheckResult.TypeCheckSuccess
+        case ArrayType(dt, _) =>
+          val dtSimple = dt.catalogString
+          TypeCheckResult.TypeCheckFailure(
+            s"$prettyName does not support sorting array of type $dtSimple which is not orderable")
+        case _ =>
+          TypeCheckResult.TypeCheckFailure(s"$prettyName only supports array input.")
+      }
+    } else {
+      checkArgumentDataTypes() match {
+        case TypeCheckResult.TypeCheckSuccess =>
+          val LambdaFunction(_, arguments, _) = function
+          if (arguments.size == 2 && function.dataType == IntegerType) {
+            TypeCheckResult.TypeCheckSuccess
+          } else {
+            TypeCheckResult.TypeCheckFailure("Return type of the given function has to be" +
+              " IntegerType")
+          }
+
+        case failure => failure
+      }
+    }
+  }
+
+  override def bind(f: (Expression, Seq[(DataType, Boolean)]) => LambdaFunction): ArraySort = {
+    val ArrayType(elementType, containsNull) = argument.dataType
+        copy(function =
+          f(function, (elementType, containsNull) :: (elementType, containsNull) :: Nil))
+  }
+
+  @transient lazy val (firstParam, secondParam) = {
+    val LambdaFunction(_, (firstParam: NamedLambdaVariable) +: tail, _) = function
+    val secondParam = tail.head.asInstanceOf[NamedLambdaVariable]
+    (firstParam, secondParam)
+  }
+
+  override def nullSafeEval(inputRow: InternalRow, argumentValue: Any): Any = {
+    val arr = argumentValue.asInstanceOf[ArrayData]
+    val f = functionForEval
+
+    @transient lazy val customComparator: Comparator[Any] = {
```

Comment:
We should move this out of `nullSafeEval` method, otherwise creates the comparator each time the `nullSafeEval` is called.